Add ReusePort config to confighttp

sinkingpoint opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This adds in a new field, `ReusePort` that, if set, sets the SO_REUSEPORT socket option on the listener port. If we're on non unix, this errors out instead as AFAIK SO_REUSEPORT isn't available. Cursory googling says that SO_REUSEADDR _might_ work, but I don't have a windows machine to test on.

<!-- Issue number if applicable -->
#### Link to tracking issue
#14046

<!--Describe what testing was performed and which tests were added.-->
#### Testing

We add a new test, `TestServerReusePort` which checks the behaviour of both setting and unsetting the ReusePort config with the expectation of an error if it isn't set and it not erroring when binding twice to the same port

sinkingpoint committed:
Add ReusePort config to confighttp

This adds in a new field, `ReusePort` that, if set, sets the
SO_REUSEPORT socket option on the listener port. If we're on non unix,
this errors out instead as AFAIK SO_REUSEPORT isn't available. Cursory
testing says that SO_REUSEADDR _might_ work, but I don't have a windows
machine to test on.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Add ReusePort config to confighttp

This adds in a new field, `ReusePort` that, if set, sets the
SO_REUSEPORT socket option on the listener port. If we're on non unix,
this errors out instead as AFAIK SO_REUSEPORT isn't available. Cursory
testing says that SO_REUSEADDR _might_ work, but I don't have a windows
machine to test on.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

axw commented on config/confighttp/listen_config_windows.go:
```diff
@@ -0,0 +1,18 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+//go:build windows
```
```suggestion
```

Nit: redundant due to the filename

axw commented:
Should we have this for configgrpc too? And non-HTTP/gRPC?

I'm wondering if we should make a breaking change to confighttp.ServerConfig to use confignet.AddrConfig, similar to configgrpc: https://github.com/open-telemetry/opentelemetry-collector/blob/c657d5d4e920c20052a12820abeef928e743c609/config/configgrpc/configgrpc.go#L176

Then we could add ReusePort to confignet.AddrConfig and have it in one place.

(Not necessarily all in this PR, just thinking out loud about how to generalise this change.)

sinkingpoint commented:
> Should we have this for configgrpc too? And non-HTTP/gRPC?

I was going to do those in a separate PR, but rely on "A little copying is better than a little dependency" and just copy the same config changes across. Breaking changes scare me

sinkingpoint committed:
ci fix

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
ci fix

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint commented:
Apologies 😬 I think I've fixed it up now anyway

axw commented:
> Actually, looking at it a bit more: configgrpc provides .Serve(net.Listener), and all the uses in the contrib repo actually use confighttps ToListener to create the listener to pass into that, so this handles both

Where did you see that being done? I found the stef receiver is not using confighttp, for example: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e3f1205a034ff05c1c31e1e1b84a8bf735563d91/receiver/stefreceiver/stef.go#L57-L77

axw commented:
Ah I see. That receiver probably should be updated, but I digress

A separate PR sounds good to me, maybe change the description so the issue doesn't get closed when this PR is merged

sinkingpoint commented:
Done. Is there anything blocking this PR then? I see the CodeCov, but those lines _are_ covered, just in the windows tests (which tracks the error flow)

sinkingpoint committed:
Add Validate method to ServerConfig

This adds a new Validate method that checks the getListenConfig call so
that the new SO_REUSEPORT stuff fails in dry run on Windows

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Add Validate method to ServerConfig

This adds a new Validate method that checks the getListenConfig call so
that the new SO_REUSEPORT stuff fails in dry run on Windows

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

dmathieu commented on config/confighttp/listen_config_windows.go:
```diff
@@ -0,0 +1,18 @@
+	}
+
+	return net.ListenConfig{}, nil
+}
```
IMHO, introducing `Validate` should be another PR.

sinkingpoint replied:
@dmathieu @bogdandrutu Where are we with this? I've added the Validate for now - IMO it _does_ make sense to have it here so there isn't a period where it _isn't_ validated. Especially with this change being the only thing in the Validate method, bundling make sense here.

dmathieu replied:
If @bogdandrutu is fine with this, it can go.

axw reviewed (approved):
@sinkingpoint sorry for the delay, LGTM. Can you please merge main?

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

sinkingpoint committed:
Merge branch 'sinkingpoint/so-reuse-port' of github.com:sinkingpoint/opentelemetry-collector into sinkingpoint/so-reuse-port

axw commented:
@sinkingpoint yes from my perspective, thanks

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

Commit:
Merge branch 'main' into sinkingpoint/so-reuse-port

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Merge branch 'sinkingpoint/so-reuse-port' of github.com:sinkingpoint/opentelemetry-collector into sinkingpoint/so-reuse-port

sinkingpoint commented:
@dmathieu @bogdandrutu Can this go now?

sinkingpoint commented:
Test failures are all the same - https://github.com/open-telemetry/opentelemetry-collector/commit/73f090cd0d59a3faa237168750d44a0036d8fd5c broke it

```
 Error: ../../internal/aws/ecsutil/client.go:65:46: cannot use host (variable of interface type component.Host) as map[component.ID]component.Component value in argument to clientSettings.ToClient
```

axw commented:
@sinkingpoint I think the best thing to do would be to update this PR to move the new config field to `confignet.AddrConfig`, and code to `confignet.AddrConfig.Listen`. That will immediately apply to configgrpc, since it already uses confignet. Once https://github.com/open-telemetry/opentelemetry-collector/pull/14248 is merged, it will also start applying to confighttp.

According to https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/release.md#release-schedule, v0.142.0 is due to be released today-ish. After that, it should be safe to merge https://github.com/open-telemetry/opentelemetry-collector/pull/14248.

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

sinkingpoint committed:
Move ReusePort to confignet

Rather than having this in confighttp specifically, this allows us to
reuse it in the net AddrConfig stuff

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Update changelog entry

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
CI fixes

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Shuffle listen_config files

As per PR comments, this moves things around so that listen_config_other
becomes listen_config_unix, and listen_config_windows becomes
listen_config_other, with appropriate build gating.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

evan-bradley reviewed (commented):
Thanks for your patience with the long review cycle @sinkingpoint.

sinkingpoint committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector into sinkingpoint/so-reuse-port

sinkingpoint committed:
Document reuse-port setting more

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint commented on config/confignet/confignet.go:
```diff
@@ -84,6 +84,13 @@ type AddrConfig struct {
+	// ReusePort enables the SO_REUSEPORT socket option on the listener.
+	// This allows multiple server instances to bind to the same address/port.
+	// This is useful for horizontal scaling and zero-downtime restarts.
+	// Note: This option is not supported on all operating systems.
```
Done

sinkingpoint committed:
Fix spell check on confignet readme

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Update ReusePort tests to properly scope checks

This changes the if runtime.GOOS == "windows" { to properly match the
updated build tags

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

evan-bradley reviewed (approved):
Thanks @sinkingpoint. I'd like to get approvals from a few other approvers/maintainers since this is a stable module and we can't revert this change once it's released.

evan-bradley commented on config/confignet/listen_config_unix.go:
```diff
@@ -0,0 +1,30 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+//go:build linux || darwin
```
We can address this later if needed.

Commit:
Add SO_REUSEPORT to cspell.json

Commit:
Fix error assertion

Commit:
Merge branch 'main' into sinkingpoint/so-reuse-port

Commit:
Merge branch 'main' into sinkingpoint/so-reuse-port
